### PR TITLE
Convert site from PSP to Jekyll

### PR DESCRIPTION
**Pull request by Indanz**
This makes the site fully static without any runtime dynamic code.

```
$ make help
Usage: make <target>. Available seL4 website targets:
	
	build       Generate the static files and put them in _site/.
	serve       Host site locally on port 4000 for previewing before commit.
	debug       Host the development version of the site for local testing.
	clean       Remove generated files.
	doctor      Run jekyll doctor to check configuration.
	checklinks  Runs html-proofer to check for broken links.
	update      Updates all Ruby Gem versions to the newest available.
	help        This help text.
```
TODO:
- Make the old validator and checking work for the new site. It currently only has a link checker.
- Remove Docker and Apache stuff or make it work.
- Update .github/* files.

### CODE REVIEW

**Comment by lsf37** [comment]
Cool, that is pretty extensive and quicker than I thought. Will need a bit of time to try out and review.

**Comment by lsf37** [comment]
Still early days, but @wom-bat if you have any input for the website conversion to Jekyll, esp from the deployment side, please let us know.

**Comment by Indanz** [comment]
Deployment should be very similar to the docs site, just copy the content of _site/ and host it somewhere.

**Comment by lsf37** [comment]
Have now looked through and I'm happy with it. Before we can merge, I'll need @wom-bat  to adjust the deployment side (should be relatively quick once @wom-bat is available), and we should also adjust the github workflows + docker image. I might have a look at the latter two today and push to the branch.

- [ ] deployment update
- [ ] docker
- [ ] adjust GitHub workflow

**Comment by lsf37** [comment]
Will rebase and add the contents of #313 when I'm finished fighting GitHub workflows..

**Comment by lsf37** [comment]
Ok, rebase is done, docker setup is also done, and I've started on the GitHub workflow (hence the many force pushes).

The GitHub checks now:
- build the site in jekyll (and fail if there is a problem)
- run the HTML 5 validator and report issues
- deploy to the preview branch and post the usual PR comment (only for in-repo branches, as before)

I've made the workflow fail a bit more gracefully for PRs from forks. As a pull_request workflow, it has insufficient access to comment on the PR and push to the preview branch when it comes from a fork. We now skip these steps (and don't fail), but still get output from the HTML 5 validator.

Issues still to solve:
- the previews are generated, but don't really work, because all links (incl to css) are root-relative instead of document relative, and the root on the preview site is not at `/` (but the production site is). Not sure if this would be easy to change in jekyll, or if I should try to generate the site for the preview with a different base URL 
- the link checker doesn't run yet. It should be possible to let it run on the site that docker serves locally, but I haven't played with that enough yet to know if it will work.

**Comment by Indanz** [comment]
> I've made the workflow fail a bit more gracefully for PRs from forks. As a pull_request workflow, it has insufficient access to comment on the PR and push to the preview branch when it comes from a fork. We now skip these steps (and don't fail), but still get output from the HTML 5 validator.

That's fine, because it's easy to do a local preview (before committing even) with "make serve" or "make debug" (only difference is Google analytics).
 
>    the previews are generated, but don't really work, because all links (incl to css) are root-relative instead of document relative, and the root on the preview site is not at `/` (but the production site is). Not sure if this would be easy to change in jekyll, or if I should try to generate the site for the preview with a different base URL

That can be fixed by using different `baseurl` in `_config`:

> To switch part of your config settings depending on the environment, use the [build command option](https://jekyllrb.com/docs/configuration/options/#build-command-options), for example --config _config.yml,_config_development.yml. Settings in later files override settings in earlier files.

So just add another _preview.yml with a different baseurl & url and run Jekyll with `--config _config.yml,_preview.yml.

> the link checker doesn't run yet. It should be possible to let it run on the site that docker serves locally, but I haven't played with that enough yet to know if it will work.

So there's "make checklinks" that does link checking with html-proofer, but it seems to give many false positives for external links. It's fine for local link checking though. It's also simplistic that it doesn't collect all links first, but checks the same link multiple times.

I think external link checking should be done periodically and not at every pull request.

File: docker/Dockerfile
Comment (Indanz) [inline docker/Dockerfile:2]: If this is based on another file somewhere else, it's worth mentioning where it comes from in the commit message (in case in the future this needs to be updated the same way as the source file).

**Comment by lsf37** [comment]
> > I've made the workflow fail a bit more gracefully for PRs from forks. As a pull_request workflow, it has insufficient access to comment on the PR and push to the preview branch when it comes from a fork. We now skip these steps (and don't fail), but still get output from the HTML 5 validator.
> 
> That's fine, because it's easy to do a local preview (before committing even) with "make serve" or "make debug" (only difference is Google analytics).

For authoring that works well, for reviewing the online preview is nice, because you don't have to check out anything. It's not highest priority, but I'll fiddle around with it a bit more this week to see if I can get it working.

> > the previews are generated, but don't really work, because all links (incl to css) are root-relative instead of document relative, and the root on the preview site is not at `/` (but the production site is). Not sure if this would be easy to change in jekyll, or if I should try to generate the site for the preview with a different base URL
> 
> That can be fixed by using different `baseurl` in `_config`:
> 
> > To switch part of your config settings depending on the environment, use the [build command option](https://jekyllrb.com/docs/configuration/options/#build-command-options), for example --config _config.yml,_config_development.yml. Settings in later files override settings in earlier files.
> 
> So just add another _preview.yml with a different baseurl & url and run Jekyll with `--config _config.yml,_preview.yml.

Thanks, that sounds promising. Will give that a go.

> > the link checker doesn't run yet. It should be possible to let it run on the site that docker serves locally, but I haven't played with that enough yet to know if it will work.
> 
> So there's "make checklinks" that does link checking with html-proofer, but it seems to give many false positives for external links. It's fine for local link checking though. It's also simplistic that it doesn't collect all links first, but checks the same link multiple times.

I should look into that a bit more.

> I think external link checking should be done periodically and not at every pull request.

I do prefer the PR checks -- firstly, because I think links should be checked when they are introduced (that's at PR time), and secondly, almost nobody checks failures of the periodic jobs, so it mostly falls back to me to fix stuff there, which I don't like :-). If there is some nagging on PRs for broken links, that can be annoying, but it tends to lead to them being fixed :-)

File: docker/Dockerfile
Comment (Indanz) [reply docker/Dockerfile:2]: > This is from scratch, I haven't based it on anything else.

Considering this work is part of the web doc revamp project, shouldn't the copyright be different? Or can I also just copyright new files in my name as long as it's open source? I assumed I had to assign everything to the seL4 foundation.

Comment (lsf37) [reply docker/Dockerfile:2]: Yes, it's Ok to keep copyright to yourself, the only requirements are the DCO and using the foundation-approved licenses. No need to assign copyright to the foundation.

**Comment by Indanz** [comment]
> For authoring that works well, for reviewing the online preview is nice, because you don't have to check out anything. It's not highest priority, but I'll fiddle around with it a bit more this week to see if I can get it working.

Good point. However, considering this PR comes from a fork, how are you going to check if the preview is working? Should we merge it to a branch first?

> I do prefer the PR checks -- firstly, because I think links should be checked when they are introduced (that's at PR time), and secondly, almost nobody checks failures of the periodic jobs, so it mostly falls back to me to fix stuff there, which I don't like :-). If there is some nagging on PRs for broken links, that can be annoying, but it tends to lead to them being fixed :-)

Fair enough! :-)

**Comment by lsf37** [comment]
> > For authoring that works well, for reviewing the online preview is nice, because you don't have to check out anything. It's not highest priority, but I'll fiddle around with it a bit more this week to see if I can get it working.
> 
> Good point. However, considering this PR comes from a fork, how are you going to check if the preview is working? Should we merge it to a branch first?

I've been testing it over at https://github.com/lsf37/website/pull/1. Probably better to do that testing separately, it tends to generate a lot of noise on the PR discussion.

**Comment by Indanz** [comment]
> I've been testing it over at [lsf37#1](https://github.com/lsf37/website/pull/1). Probably better to do that testing separately, it tends to generate a lot of noise on the PR discussion.

Okay if I push more commits to my fork? I want to give a stab at converting news to posts, but don't want to make your life more difficult.

**Comment by lsf37** [comment]
> > I've been testing it over at [lsf37#1](https://github.com/lsf37/website/pull/1). Probably better to do that testing separately, it tends to generate a lot of noise on the PR discussion.
> 
> Okay if I push more commits to my fork? I want to give a stab at converting news to posts, but don't want to make your life more difficult.

Yes, no problem.

In a different setup we used a data file for news with items "date", "title", and "content" where the content is markdown (in that setup it's toml, because it is done in hugo, but yml would probably work). We then iterate over that in the templates and generate one news page per year plus clippings of the most recent three on the title page. Not sure if that is better than posts or not, but I thought I mention the idea in case posts don't work out nicely.

**Comment by Indanz** [comment]
> In a different setup we used a data file for news with items "date", "title", and "content" where the content is markdown (in that setup it's toml, because it is done in hugo, but yml would probably work). We then iterate over that in the templates and generate one news page per year plus clippings of the most recent three on the title page. Not sure if that is better than posts or not, but I thought I mention the idea in case posts don't work out nicely.

That's more or less what I wanted to do with posts. Posts are more flexible though, they can be either markdown or html. I want to make it as easy as possible to add a common news item with one image floating on the right, but because html is also supported, it's still possible to add more images or do other formatting if wanted. I'll keep most of the current news items as html and only convert one or two to markdown as an example.

**Comment by lsf37** [comment]
Apologies for the force push in between. The link checker now uses the Makefile `checklinks` target and appears to be working fine (manually ignoring the rate-limited twitter links).

**Comment by Indanz** [comment]
> Apologies for the force push in between. The link checker now uses the Makefile `checklinks` target and appears to be working fine (manually ignoring the rate-limited twitter links).

No problem. Looks good to me.

**Comment by lsf37** [comment]
So I played around a bit more with the GitHub preview feature, and there are complications. Setting `site` and `baseurl` in the config only has an effect if the corresponding URL is filtered through something like `relative_url` or `absolute_url` in liquid. I.e. we'd have to turn every single `href a=".."` into something like `href a={{ "..." | relative_url }}` which is both messy and likely to be forgotten in new contributions.

It would be easy enough to do in things that are generated anyway (menu, breadcrumbs, templates, etc), but there are many links in the content part. Not sure if there is a way to hook into the `md` to `html` conversion, but we don't have a lot of `md` yet anyway. The only other option I can think of is that we try to stick to relative URLs in the content, i.e. instead of linking to `/images/pic.jpg` we use `images/pic.jpg` or `../images/pic.jpg` etc. Of course that then depends on which directory you're in.

**Comment by Indanz** [comment]
> So I played around a bit more with the GitHub preview feature, and there are complications. Setting `site` and `baseurl` in the config only has an effect if the corresponding URL is filtered through something like `relative_url` or `absolute_url` in liquid. I.e. we'd have to turn every single `href a=".."` into something like `href a={{ "..." | relative_url }}` which is both messy and likely to be forgotten in new contributions.
> 
> It would be easy enough to do in things that are generated anyway (menu, breadcrumbs, templates, etc), but there are many links in the content part. Not sure if there is a way to hook into the `md` to `html` conversion, but we don't have a lot of `md` yet anyway. The only other option I can think of is that we try to stick to relative URLs in the content, i.e. instead of linking to `/images/pic.jpg` we use `images/pic.jpg` or `../images/pic.jpg` etc. Of course that then depends on which directory you're in.

If we support `baseurl`, links in templates should use the filters and the other URLs should be relative, generally speaking.

How is this problem solved with the current preview setup? That should have the same issue.

I'll bring this branch up-to-date later this week.

**Comment by lsf37** [comment]
> If we support `baseurl`, links in templates should use the filters and the other URLs should be relative, generally speaking.

Ok, happy with this. Let's do that.
 
> How is this problem solved with the current preview setup? That should have the same issue.

It uses a more elaborate setup to get the preview snapshot: it hosts the dynamic site on apache inside docker, and then crawls it with `wget` to get a static snapshot. It uses the `wget` option `--convert-links` to change all links to local relative links. We could in theory still do the same, but it sounds like overkill. Better to use relative links directly and filtered links in the templates.

> I'll bring this branch up-to-date later this week.

👍

**Comment by wom-bat** [comment]
So I'm starting to look at this.  It has some conflicts that need fixing before merging;  but overall looks good.
I want to see if I can get the modified site running locally --- I'm not convinced for instance that the <title> tags will be correctly generated as they've been deleted from the templating; but I don't really understand Jekyll.

**Comment by wom-bat** [comment]
OK Titles are good.

**Comment by wom-bat** [comment]
This still has google analytics tracking, although it's correctly disabled for browsers with 'do not track' set

**Comment by wom-bat** [comment]
OK I think this is mergeworthy.   There are minor merges that we'll need to fix, but that should be almost trivial.
I'm currently checking this can be served from seL4/LionsOS, but I don't foresee any issues.

**Comment by wom-bat** [comment]
One thing I notice: the copyright and SPDX identifiers that used to be visible in HTML comments in the webage if you chose to look at the source are no longer there.   They're still there as Jekyll comments in the source on github, but aren't delivered to the end user.  Is this deliberate?

**Comment by Indanz** [comment]
> This still has google analytics tracking, although it's correctly disabled for browsers with 'do not track' set

I retained that because I assumed that was wanted. The original site had it. The tracking should be disabled if you do `make debug` to host and test it locally.

> One thing I notice: the copyright and SPDX identifiers that used to be visible in HTML comments in the webage if you chose to look at the source are no longer there. They're still there as Jekyll comments in the source on github, but aren't delivered to the end user. Is this deliberate?

No, I just didn't think about that when converting. If we do want user visible copyrights, it should be done once in the template I think, not on each individual page. But whatever you prefer.

**Review by wom-bat — APPROVED** [review]
There are changes to make during the merge (to bring up to date with the master site) but this looks OK.  I don't foresee any problems in deploying it.

File: .github/workflows/main.yml
Comment (lsf37) [reply .github/workflows/main.yml:53]: I do find the two separate invocations slightly easier to read, but the real reason is that this is just what Luke had in his version previously. Happy to merge it in if you like.

File: Foundation/Services/endorsement.html
Comment (Indanz) [reply Foundation/Services/endorsement.html:56]: Other than fixing obvious bugs, I tried minimising unrelated changes and improvements.

Agreed the site could use a good clean over, but that should happen in a separate PR. This one is hard enough to review a it is.

File: Foundation/Services/index.html
Comment (Indanz) [reply Foundation/Services/index.html:22]: As far as I know, that's not standard HTTP and would rely on a specific Apache setting or module to make it work.

File: _layouts/default.html
Comment (Indanz) [reply _layouts/default.html:48]: I'm happy to do that, that would simplify things a bit, but it was added deliberately to the original website, so I assumed seL4 has a Google account and someone keeps tracks of the statistics.

Comment (lsf37) [reply _layouts/default.html:48]: We checked -- we used to, but it's not used any more and to activate it we would have to add a GDPR pop-up, which we don't want. So Peter is right, we should remove the whole thing.

File: Foundation/Services/index.html
Comment (wom-bat) [reply Foundation/Services/index.html:22]: It's standard on all the webservers we use.  Unadorned filenames are looked up, then tried again with a '.html' suffix.  For the python version, '.pml' was always optional.

**Comment by Indanz** [comment]
TODOs:

- Visible copyrights per page or once in template? Now it's not consistent.
- Support `baseurl` and fix docker preview by making all urls relative. This could also be done after merging with master, to avoid future tricky merges.
- Convert news to posts. Also something for post-merge.

So I think it's almost ready for merging, let me know if I missed anything.

**Comment by Indanz** [comment]
> It looks like I should switch off the benchmark update script for a bit until we have merged this, because it has created a conflict already again.

At least it automatically merges, so it's quick to fix.

> The site is now online on GitHub pages under https://sel4.github.io/website/ -- it doesn't work properly because the links for all resources, including images, are all absolute (go from the root), and it's currently serving under a folder (same as with the preview). This will disappear once we serve it under `sel4.systems/` directly.

Can't you get it to run under sel4.github.io/ directly? That should fix it too and there isn't anything else at the root yet.

I did this, results here (with modified _preview.yml): https://indanz.github.io/

**Comment by lsf37** [comment]
> > It looks like I should switch off the benchmark update script for a bit until we have merged this, because it has created a conflict already again.
> 
> At least it automatically merges, so it's quick to fix.

For me it's showing "cannot rebase because of conflicts".


> > The site is now online on GitHub pages under https://sel4.github.io/website/ -- it doesn't work properly because the links for all resources, including images, are all absolute (go from the root), and it's currently serving under a folder (same as with the preview). This will disappear once we serve it under `sel4.systems/` directly.
> 
> Can't you get it to run under sel4.github.io/ directly? That should fix it too and there isn't anything else at the root yet.

No that doesn't work -- for PR previews you need a separate site view per PR, because there can be multiple at the same time. The only way I can see this working is with a sub path with some ID like the PR number in it (the current preview does that). For that it doesn't really matter if it's our own GH pages or the preview service it is currently using. We could use GH pages on the existing seL4 preview repo for it, that would work, but the issue with the links being absolute doesn't go away.

> I did this, results here (with modified _preview.yml): https://indanz.github.io/

This would only work for a single PR.

**Comment by lsf37** [comment]
I'm happy to leave the commit with the `preview` target in, because it is a necessary step for it to work later. It just needs additional ones.

**Comment by Indanz** [comment]
> > > It looks like I should switch off the benchmark update script for a bit until we have merged this, because it has created a conflict already again.
> > 
> > 
> > At least it automatically merges, so it's quick to fix.
> 
> For me it's showing "cannot rebase because of conflicts".

You can't rebase it via the github interface, but when doing `git merge` locally it automatically merges fine.

> No that doesn't work -- for PR previews you need a separate site view per PR, because there can be multiple at the same time. The only way I can see this working is with a sub path with some ID like the PR number in it (the current preview does that). For that it doesn't really matter if it's our own GH pages or the preview service it is currently using. We could use GH pages on the existing seL4 preview repo for it, that would work, but the issue with the links being absolute doesn't go away.
> 
> > I did this, results here (with modified _preview.yml): https://indanz.github.io/
> 
> This would only work for a single PR.

Okay, I might have misnamed the `preview` target, it wasn't intended for PR previews, more like GH pages export.

For PR previews we have to do what I mentioned in the TODO list above, support baseurl and make all urls relative, but this is something I want to do post-merge of this PR.

**Comment by lsf37** [comment]
> > > > It looks like I should switch off the benchmark update script for a bit until we have merged this, because it has created a conflict already again.
> > > 
> > > 
> > > At least it automatically merges, so it's quick to fix.
> > 
> > 
> > For me it's showing "cannot rebase because of conflicts".
> 
> You can't rebase it via the github interface, but when doing `git merge` locally it automatically merges fine.

Ok, good. We still need to rebase to actually merge the PR, the master branch doesn't allow merge commits.


> Okay, I might have misnamed the `preview` target, it wasn't intended for PR previews, more like GH pages export.
> 
> For PR previews we have to do what I mentioned in the TODO list above, support baseurl and make all urls relative, but this is something I want to do post-merge of this PR.

Agreed. From my side, we're ready to merge this PR.

**Review by lsf37 — APPROVED** [review]

**Comment by Indanz** [comment]
> Ok, good. We still need to rebase to actually merge the PR, the master branch doesn't allow merge commits.

Ah, yes, that will be fun. I'll do the rebase in a separate checkout so I can double check the content is the same as for the current state of the PR. This might take a few hours.

**Comment by lsf37** [comment]
The deployment workflow is also ready. Commit https://github.com/seL4/website/commit/bb3dad51f7e0fb48aae8e742bd4d72bbfd6d8132 on branch [auto-deploy](https://github.com/seL4/website/tree/auto-deploy) could be pulled into this PR if you like, but can also added after merging this. Whatever is easier.

**Comment by Indanz** [comment]
Okay, doing the rebase was kind of hell. Basically I had to fix all merge conflict again, so all the intermediate merges to bring it up-to-date with master were a waste of time in hindsight.

I lumped the merge fixes with more or less random commits, so it doesn't always make sense, but it ended up where it caused a conflict.

I did a diff between the pre-rebase and this version, the changes are reduces to details. The diff is:

```Only in ../website-rebase/: .jekyll-cache
Only in ../website-rebase/Foundation/Board: 230928-minutes-unconfirmed.pdf
Only in ./Foundation/Board/Minutes: 230928-minutes-unconfirmed.pdf
diff -pru '--exclude=.git' '--exclude=_preview' '--exclude=vendor' ./Foundation/Membership/index.html ../website-rebase/Foundation/Membership/index.html
--- ./Foundation/Membership/index.html	2024-03-19 16:00:41.705202367 +0000
+++ ../website-rebase/Foundation/Membership/index.html	2024-03-20 09:52:13.199480778 +0000
@@ -220,7 +220,7 @@ redirect_from: /Foundation/Membership/ho
   <div class="members">
     <div class="member">
       <div class="member-pic">
-        <a href="https://autoware.org/">
+        <a href="https://autoware.org">
           <img src="LOGOS/autoware.svg" style="width: 200px" alt="Autoware Foundation">
         </a>
       </div>
diff -pru '--exclude=.git' '--exclude=_preview' '--exclude=vendor' ./Foundation/Summit/2022/principles.html ../website-rebase/Foundation/Summit/2022/principles.html
--- ./Foundation/Summit/2022/principles.html	2024-03-19 16:00:41.706202376 +0000
+++ ../website-rebase/Foundation/Summit/2022/principles.html	2024-03-20 09:52:13.210480843 +0000
@@ -1,10 +1,8 @@
 ---
+# Copyright 2022 seL4 Project a Series of LF Projects, LLC.
+# SPDX-License-Identifier: CC-BY-SA-4.0
 title: Principles
 ---
-<!--
-    Copyright 2023 seL4 Project a Series of LF Projects, LLC.
-    SPDX-License-Identifier: CC-BY-SA-4.0
--->
 <h1>
   seL4 Summit
 </h1>
diff -pru '--exclude=.git' '--exclude=_preview' '--exclude=vendor' ./Foundation/Summit/index.html ../website-rebase/Foundation/Summit/index.html
--- ./Foundation/Summit/index.html	2024-03-19 16:00:42.939212897 +0000
+++ ../website-rebase/Foundation/Summit/index.html	2024-03-20 09:52:13.266481170 +0000
@@ -1,5 +1,5 @@
 ---
-# Copyright 2024 seL4 Project a Series of LF Projects, LLC.
+# Copyright 2023 seL4 Project a Series of LF Projects, LLC.
 # SPDX-License-Identifier: CC-BY-SA-4.0
 title: The seL4 Summit
 redirect_from: /Foundation/Summit/home.pml
diff -pru '--exclude=.git' '--exclude=_preview' '--exclude=vendor' ./_layouts/default.html ../website-rebase/_layouts/default.html
--- ./_layouts/default.html	2024-03-19 16:00:43.059213921 +0000
+++ ../website-rebase/_layouts/default.html	2024-03-20 09:52:14.280487088 +0000
@@ -80,7 +80,7 @@
       <a href="https://www.linkedin.com/company/sel4"><img style="height:36px; padding-left:5px; padding-top:2px; padding-bottom:2px;" src="/images/icons/linkedin.svg" alt="linkedin icon"></a>
       <br>
       <br>
-    Copyright &copy; 2023 seL4 Project a Series of LF Projects, LLC.<br/>
+    Copyright &copy; 2024 seL4 Project a Series of LF Projects, LLC.<br/>
       seL4 is a trademark of LF Projects, LLC.<br/>
       For trademark usage guidelines, privacy and cookie policies,
       and other applicable policies, as well as terms and conditions
@@ -91,7 +91,7 @@
     <div class="row">
       <!-- for screens larger than mobiles -->
       <div class="col-xs-6 hidden-xs">
-        Copyright &copy; 2023 seL4 Project a Series of LF Projects, LLC.<br/>
+        Copyright &copy; 2024 seL4 Project a Series of LF Projects, LLC.<br/>
         seL4 is a trademark of LF Projects, LLC.<br/>
         For trademark usage guidelines, privacy and cookie policies,
         and other applicable policies, as well as terms and conditions
Only in ../website-rebase/: _site
Only in .: rebase.diff
```

**Pull request merged**
